### PR TITLE
fix: allow safe retry before provider progress

### LINKED ISSUE #912: [Bug] Safe retry is blocked by unknown exposed tool boundaries before provider progress
## What happened?

A production build still showed the conservative interruption message:

> The connection was interrupted, and PawWork could not prove whether external side effects were possible.

The session export shows the terminal request failed with an OpenAI `Service Unavailable` before first provider progress. No assistant output started, no tool input started, no tool call materialized, and no tool execution started. However, incident recovery still downgraded the run to `ask_user_before_retry` because the global side-effect boundary proof was incomplete: 10 exposed tools were classified as unknown.

This means a run that appears dynamically safe to retry is blocked by static unknown exposed-tool metadata, even when none of those tools were called.

## Which area seems affected?

Model harness, prompts, tools, or session mechanics

## How much does this affect you?

Makes a workflow harder, but there is a workaround

## Steps to reproduce

1. Open PawWork production build `0.0.0-prod-202605251621`.
2. Start a session with the normal tool set exposed.
3. Hit a provider transport failure before first provider progress, such as OpenAI `Service Unavailable`.
4. Inspect the session export / incident diagnostics.
5. Observe that the terminal attempt has no visible output and no tool activity, but recovery still asks the user before retrying because `side_effect_facts_complete` is false.

Observed export facts from `pawwork-session-mighty-cactus-2026-05-26-01-33-38.json`:

```json
{
  "classification": "external_stream_disconnect",
  "summary_key": "external_stream_disconnect.transport_failure",
  "provider_progress_seen": false,
  "visible_output_seen": false,
  "tool_input_started": false,
  "tool_call_materialized": false,
  "tool_execution_started": false,
  "unsafe_side_effect_started": false,
  "retry_safety": {
    "recommendation": "candidate_safe_auto_retry",
    "reason": "no_visible_output_or_tool_execution"
  },
  "side_effect_facts_complete": false,
  "side_effect_boundary_snapshot": {
    "exposed_tool_count": 16,
    "unknown_tool_count": 10,
    "proof_result": "incomplete",
    "proof_reason": "unknown_tool_boundary"
  },
  "incident": {
    "recovery": {
      "recommendation": "ask_user_before_retry",
      "reason": "side_effect_facts_incomplete"
    },
    "missing_provenance": ["side_effect.boundary_unknown"]
  }
}
```

Relevant code path:

- `packages/opencode/src/session/processor.ts`: `sideEffectBoundarySnapshot()` marks the exposed tool set incomplete when any exposed tool effect is unknown.
- `packages/opencode/src/session/run-incident/policy.ts`: `!terminalFacts.side_effect_facts_complete` returns `ask_user_before_retry` before the later retryable-transport auto-retry path can run.
- `packages/opencode/src/session/processor.ts`: `side_effect_facts_incomplete` maps to the conservative user-facing interruption message.

## What did you expect to happen?

If the terminal attempt failed before first provider progress and no output/tool activity occurred, PawWork should auto-retry once or show the safer early-transport-failure message. Unknown exposed tools should not block safe retry when no tool input started, no tool call materialized, and no tool execution started.

## PawWork version

`0.0.0-prod-202605251621`

## OS version

macOS `25.5.0`

## Can you reproduce it again?

Only once so far

## Diagnostics

Session export inspected locally:

`/Users/yuhan/Downloads/pawwork-session-mighty-cactus-2026-05-26-01-33-38.json`

Potential fix directions:

- Minimal: add an early recovery fast path for retryable transport failures before first provider progress with no visible output and no tool activity, bypassing incomplete exposed-tool boundary proof unless provider-executed/external boundaries are present.
- Cleaner follow-up: separate static offered-tool boundary completeness from dynamic attempt-side-effect possibility, so recovery policy is based on what actually happened before interruption.

### PR DESCRIPTION
**Pull request by Astro-Han**
## Summary

- Allow one automatic retry for retryable provider transport failures that happen before first provider progress when the terminal attempt produced no output and no tool activity.
- Keep the path fail-closed when the boundary snapshot is missing, provider/external boundaries are present, lifecycle/user cancel is observed, or telemetry shows any output/tool activity.
- Add regression coverage for the Issue #912 export shape plus negative cases for missing evidence and contradictory tool/output state.

## Why

Issue #912 showed a production session where the provider returned `Service Unavailable` before any provider progress, assistant output, tool input, tool call, or tool execution. PawWork still asked the user before retrying because unknown exposed tool metadata made the static side-effect boundary proof incomplete. Retry safety should be based on what the terminal attempt actually reached; unknown metadata for uncalled tools should not block this early safe-retry case.

## Related Issue

Closes #912

## Human Review Status

- `Pending` — waiting for a human reviewer to approve.

## Review Focus

- The new `canAutoRetryBeforeFirstProviderProgress` predicate in `packages/opencode/src/session/run-incident/policy.ts`, especially the fail-closed guards for missing snapshots, global lifecycle/user cancel, provider/external boundaries, and contradictory output/tool facts.
- The negative matrix in `packages/opencode/test/session/run-observability.test.ts` to ensure the fast path only applies to truly early no-output/no-tool failures.

## Risk Notes

The behavior change is intentionally narrow, but it affects retry policy for model transport failures. The main risk is over-retrying if evidence is incomplete; the fast path requires a recorded boundary snapshot and explicit absence of provider/external boundaries to mitigate that.

Skipped conditional checklist items:

- Visible UI/copy check: not applicable; no visible UI or copy changed.
- Platform impact check: not applicable; no platform, packaging, updater, signing, paths, shell, or permissions surface changed.
- Docs/release/dependencies/etc.: not applicable; no docs, release notes, dependencies, permissions, credentials, generated content, or local file behavior changed.

## How To Verify

```text
RED check: bun test test/session/run-observability.test.ts failed on the new Issue #912 regression with side_effect_facts_incomplete before the fix.
Focused observability tests: bun test test/session/run-observability.test.ts — 62 passed.
Processor integration tests: bun test test/session/processor-effect.test.ts — 29 passed.
Typecheck: bun run typecheck in packages/opencode — passed.
Diff check: git diff --check — no whitespace errors.
```

## Screenshots or Recordings

Not applicable; no visible UI changes.

## Checklist

> **How to use this checklist:**
>
> - Tick a box by replacing `[ ]` with `[x]`. Do not edit, add, or remove items.
> - The bot-applied label items can only be honestly ticked AFTER the PR is opened and the labeler / priority-triage bots have run — return to the PR description and tick them then.
> - Most items are required. The few that are conditional are explicitly marked **(conditional)**; for those, leave unticked if they truly do not apply and explain why in Risk Notes. All other items must be ticked before requesting human review.

- [x] **Type label** — this PR carries exactly one of `bug`, `enhancement`, `task`, `documentation`. Type labels are author-added; the labeler bot does NOT assign them. Add the label in the GitHub UI, then tick this.
- [x] **Routing labels** — this PR carries at least one of `app`, `ui`, `platform`, `harness`, `ci`. The labeler bot assigns these on PR open based on changed paths. Confirm the bot's choice (or override if wrong), then tick this.
- [x] **Priority label** — this PR carries exactly one of `P0`, `P1`, `P2`, `P3`. The priority-triage bot suggests one on PR open. Confirm or override, then tick this.
- [x] Human Review Status above is set to `Pending`, `Approved by @<reviewer>`, or `Not required: <reason>` (default is `Pending`; "not required" is restricted to bot-authored low-risk PRs).
- [x] I linked the related issue, or stated in Summary why there is no issue.
- [x] I described the review focus and any meaningful risks.
- [x] I replaced the example block in How To Verify with the real verification steps and the key result for each.
- [x] I did not introduce unrelated refactors, dependencies, generated files, or file changes beyond the stated scope.
- [ ] **(conditional)** I manually checked visible UI or copy changes when needed, with screenshots or recordings. Leave unticked only if no visible UI or copy changed.
- [ ] **(conditional)** I considered macOS and Windows impact for platform, packaging, updater, signing, paths, shell, or permissions changes. Leave unticked only if no platform/packaging surface was touched.
- [ ] **(conditional)** I called out docs, release notes, dependencies, permissions, credentials, deletion behavior, generated content, or local file changes when relevant. Leave unticked only if none of those surfaces was touched.
- [x] I reviewed the final diff for unrelated changes and suspicious dependency changes.
- [x] I am targeting `dev`, and my PR title and commit messages use Conventional Commits in English.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved error recovery logic with enhanced automatic retry capabilities during incident handling.
  * Refined safety checks to ensure retries are attempted only when appropriate conditions are met.

* **Tests**
  * Expanded test coverage for error recovery scenarios and retry behavior validation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Astro-Han/pawwork/pull/914?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/opencode/src/session/run-incident/policy.ts [hunk 1]
```diff
@@ -9,15 +9,41 @@ export function recoveryFor(input: {
...
+      confidence: "high",
+      reason: "no_visible_output_or_tool_execution",
+      auto_retry: { max_attempts: 1, backoff_ms: 1_000 },
+    }
+  }
   if (
     retryableTransport &&
     noToolActivity &&
+    !isBeforeFirstProviderProgressCause(input.cause) &&
     terminalFacts.reasoning_output_started &&
```
[the remark this answers is not in the document]
Comment (Astro-Han) [reply packages/opencode/src/session/run-incident/policy.ts:55]: Fixed in adfd11da55 — before-progress boundary evidence now blocks fallback to the generic transport auto-retry path, and the test matrix covers provider-executed/external-boundary cases with complete facts.
